weedcoco/repo/deposit.py

```diff
@@ -2,178 +2,362 @@
-    image_hash = create_image_hash(image_dir)
-    validate_duplicate_images(image_hash)
-    validate_existing_images(repository_dir, image_hash)
-    with tempfile.TemporaryDirectory() as temp_dir:
-        temp_dir = pathlib.Path(temp_dir)
-        dataset_dir, deposit_id = setup_dataset_dir(repository_dir, temp_dir, upload_id)
-        deposit_weedcoco(weedcoco_path, dataset_dir, image_dir, image_hash)
-        migrate_images(dataset_dir, image_dir, image_hash)
-        compress_to_download(dataset_dir, deposit_id, temp_dir)
-        atomic_copy(repository_dir, download_dir, temp_dir, deposit_id)
-    return str(repository_dir)
+        Returns the id of the object.
+        """
+        if identifier is None:
+            identifier = str(uuid4())
+            dataset.identifier = identifier
+        else:
+            assert "/" not in identifier
+        dataset.validate()
+        ocfl_metadata = ocfl.VersionMetadata(
+            identifier=identifier,
+            message=metadata["message"],
+            address=metadata["address"],
+            name=metadata["name"],
+        )
+        last_version = dataset.head_version
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_dir = pathlib.Path(temp_dir)
+            dataset_dir = self.setup_deposit(temp_dir, identifier)
+            dataset.build(dataset_dir)
+            zip_file = dataset.compress_to_download(temp_dir)
+            try:
+                if dataset.exists_in_repo:
+                    dataset.ocfl.update(
+                        objdir=str(dataset.object_path),
+                        srcdir=str(dataset_dir),
+                        metadata=ocfl_metadata,
+                    )
+                else:
+                    new_object_dir = temp_dir / pathlib.Path(identifier + "_ocfl")
+                    new_object = ocfl.Object(identifier=identifier)
+                    new_object.create(
+                        objdir=str(new_object_dir),
+                        srcdir=str(dataset_dir),
+                        metadata=ocfl_metadata,
+                    )
+                    self.ocfl.add(object_path=str(new_object_dir))
+                if zip_file.is_file():
+                    old_zip = download_dir / zip_file.name
+                    if old_zip.is_file():
+                        old_zip.unlink()
+                    move(str(zip_file), str(download_dir))
+            except Exception:
+                dataset.rollback(dataset_dir, last_version)
+                zip_file = download_dir / zip_file.name
+                if zip_file.is_file():
+                    zip_file.unlink()  # missing_ok is not in 3.7
+                raise
+        return dataset
+
+
+def deposit(
+    weedcoco_path, image_dir, repository_dir, download_dir, metadata, upload_id=None
+):
+    repository = Repository(repository_dir)
+    repository.initialize()
+    dataset = repository.dataset(upload_id)
```

Comment:
I'm thinking how to return an existing database with the upload_id, to serve the editing_dataset method.

Reply:
As it is now, I think too much of the code for atomicity is in the Repository class instead of the RepositoryDataset, so adapting it for updates will be a bit clunky.

It would be better, maybe, if it worked like this - whether upload_id exists or not:

```
repository = Repository(repository_dir)
dataset = repository.dataset(upload_id)
with dataset.checkin_version() as tmp_dir:
    # code which builds a dataset in tmp_dir
```

And the `with` context would take care of rolling things back etc